lxd/storage/drivers/driver_dir.go

```diff
@@ -0,0 +1,293 @@
+	}
+
+	// Unmount until nothing is left mounted.
+	return forceUnmount(path)
+}
+
+func (d *dir) GetResources() (*api.ResourcesStoragePool, error) {
+	// Use the generic VFS resources.
+	return vfsResources(GetPoolMountPoint(d.name))
+}
+
+// DeleteVolume deletes a volume of the storage device. If the volume supplied is a snapshot and
+// it is the last one being removed for the parent volume then the snapshots directory for that
+// parent volume is removed too.
+func (d *dir) DeleteVolume(volType VolumeType, volName string, op *operations.Operation) error {
+	volPath := GetVolumeMountPoint(d.name, volType, volName)
+
+	if !shared.PathExists(volPath) {
+		return nil
+	}
+
+	volID, err := d.getVolID(volType, volName)
+	if err != nil {
+		return err
+	}
+
+	err = d.deleteQuota(volPath, volID)
+	if err != nil {
+		return err
+	}
+
+	err = os.RemoveAll(volPath)
+	if err != nil {
+		return err
+	}
+
+	// If the volume being removed is a snapshot, check if it is the last one being removed for
+	// the volume parent and if so then remove the parent volume's snaphots directory.
+	if shared.IsSnapshot(volName) {
+		// Extract just the snapshot name from the volume name and then remove that suffix
+		// from the volume path. This will get us the parent snapshots directory we need.
+		snapName := shared.ExtractSnapshotName(volName)
+		snapshotsPath := strings.TrimSuffix(volPath, snapName)
+		isEmpty, err := shared.PathIsEmpty(snapshotsPath)
+		if err != nil {
+			return err
+		}
+
+		if isEmpty {
+			err := os.Remove(snapshotsPath)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (d *dir) RenameVolume(volType VolumeType, volName string, newVolName string, op *operations.Operation) error {
+	return ErrNotImplemented
+}
+
+func (d *dir) MigrationType() migration.MigrationFSType {
+	return migration.MigrationFSType_RSYNC
+}
+
+func (d *dir) PreservesInodes() bool {
+	return false
+}
+
+// CreateVolume creates an empty volume and can optionally fill it by executing the supplied
+// filler function.
+func (d *dir) CreateVolume(volType VolumeType, contentType ContentType, volName string, filler func(path string) error, op *operations.Operation) error {
+	if contentType != ContentTypeFS {
+		return fmt.Errorf("Content type not supported")
+	}
+
+	volPath := GetVolumeMountPoint(d.name, volType, volName)
+	err := os.MkdirAll(volPath, 0711)
+	if err != nil {
+		return err
+	}
+
+	volID, err := d.getVolID(volType, volName)
+	if err != nil {
+		os.RemoveAll(volPath)
+		return err
+	}
+
+	err = d.initQuota(volPath, volID)
+	if err != nil {
+		os.RemoveAll(volPath)
+		return err
+	}
+
+	if filler != nil {
+		err = filler(volPath)
+		if err != nil {
+			os.RemoveAll(volPath)
+			return err
+		}
+	}
+
+	return nil
+}
+
+// CreateVolumeFromCopy provides same-pool volume copying functionality.
+func (d *dir) CreateVolumeFromCopy(volType VolumeType, volName string, srcVolType VolumeType, srcVolName string, op *operations.Operation) error {
```

Comment:
I think we'll need an extra `snapshots bool` to control whether we copy snapshots or not.

Reply:
@stgraber this function is only called to copy a single low level volume (either the parent or a single snapshot), so I'm not clear what the snapshots boolean would do?

Reply:
Most drivers will not let you copy individual snapshots.

So you effectively have the following options:
 - Copy a volume and all its snapshots
 - Copy a volume and none of its snapshots
 - Copy a volume snapshot as a new volume

You can't copy a volume with no snapshot and then decide to copy the snapshots later.

Copying subset of snapshots is also unlikely to be consistently supported and isn't something we currently allow in the API, so a snapshot slice isn't really helpful.

Reply:
As for the copy from snapshot case. Since we want to be explicit about snapshot use, I think we'll want a `srcVolSnapshot string` argument. If it's empty, we copy a whole volume with or without its snapshots (depending on `snapshots bool`). If it's not empty, then we're copying from a volume snapshot into a new volume (`snapshots bool` is meaningless in that case).

Reply:
@stgraber said in IRC:

"

[the drivers] all of them have some way of listing snapshots
--
21:30:50 | <stgraber> | tomponline: we'll need to skip internal snapshots (those we create during copy and migration for zfs and ceph) but we had that kind of logic before too

for dir/btrfs/cephfs, it's just a directory lookup, for zfs/ceph/lvm, the internal commands let you  get the list
"

Reply:
@stgraber I've updated dir.CreateVolumeFromCopy and lxdBackend.CreateCustomVolumeFromCopy as we discussed. The only change I made was to change the `srcVolSnapshot` arg to `srcIsSnapshot` boolean instead. This seemed to achieve both the explicit notification to the driver that the source was a snapshot without the need to pass the same volume name (i.e volume/snapshot) in srcVolName and srcVolSnapshot.

I've used the srcIsSnapshot argument to detect when to not try and copy snapshots, even if copySnapshots is true.